File: src/lib/i18n/audit_i18n.py
"""
i18n Audit Script for TraderLogPro
Finds all $t('...') and $t("...") keys used in .svelte and .ts files
and compares them against the pt-BR locale JSONs.
"""

import json
import os
import re
import sys
from pathlib import Path
from collections import defaultdict

BASE_DIR = Path("c:/PROJETOS/TraderLogPro")
SRC_DIR = BASE_DIR / "src"
LOCALE_DIR = BASE_DIR / "src/lib/i18n/locales/pt-BR"


def load_all_ptbr_keys():
    """Load all keys from all pt-BR JSON files, flattening nested structures."""
    all_keys = {}
    for json_file in LOCALE_DIR.glob("*.json"):
        with open(json_file, encoding="utf-8") as f:
            data = json.load(f)
        flat = flatten_keys(data)
        for key, value in flat.items():
            all_keys[key] = {"value": value, "file": json_file.name}
    return all_keys


def flatten_keys(data, prefix=""):
    """Recursively flatten nested JSON keys into dot notation."""
    result = {}
    for k, v in data.items():
        full_key = f"{prefix}.{k}" if prefix else k
        if isinstance(v, dict):
            result.update(flatten_keys(v, full_key))
        else:
            result[full_key] = v
    return result


def extract_t_keys_from_file(filepath):
    """Extract all $t('key') and $t("key") calls from a file."""
    with open(filepath, encoding="utf-8", errors="ignore") as f:
        content = f.read()
    
    # Match $t('...') and $t("...") patterns
    # Also matches t('...') used in .ts files
    patterns = [
        r'\$t\(["\']([^"\')\s]+)["\']',
        r'(?<!\$)(?<!\w)t\(["\']([^"\')\s]+)["\']',
    ]
    
    keys = set()
    for pattern in patterns:
        matches = re.findall(pattern, content)
        for m in matches:
            # Skip interpolation patterns like $t(`...${...}...`)
            if "${" not in m and "{" not in m:
                keys.add(m)
    return keys


def scan_all_source_keys():
    """Scan all .svelte and .ts files for $t() keys."""
    used_keys = defaultdict(list)
    
    extensions = [".svelte", ".ts"]
    exclude_dirs = {"node_modules", ".svelte-kit", "build", "dist", "__tests__"}
    
    for ext in extensions:
        for filepath in SRC_DIR.rglob(f"*{ext}"):
            # Skip excluded directories
            if any(excluded in filepath.parts for excluded in exclude_dirs):
                continue
            # Skip test files and the i18n files themselves
            if "test" in filepath.name.lower() or "i18n" in str(filepath):
                continue
            
            keys = extract_t_keys_from_file(filepath)
            rel_path = filepath.relative_to(BASE_DIR)
            for key in keys:
                used_keys[key].append(str(rel_path))
    
    return used_keys


def main():
    print("=" * 70)
    print("TraderLogPro — i18n Audit Report")
    print("=" * 70)
    
    print("\n📂 Loading pt-BR locale keys...")
    ptbr_keys = load_all_ptbr_keys()
    print(f"   ✅ {len(ptbr_keys)} keys loaded from {LOCALE_DIR}")
    
    print("\n🔍 Scanning source files for $t() usage...")
    used_keys = scan_all_source_keys()
    print(f"   ✅ {len(used_keys)} unique keys found in source code")
    
    # Find missing keys (used but not in pt-BR)
    missing = {k: v for k, v in used_keys.items() if k not in ptbr_keys}
    
    # Find unused keys (in pt-BR but not used in code)
    used_set = set(used_keys.keys())
    ptbr_set = set(ptbr_keys.keys())
    unused = {k: ptbr_keys[k] for k in ptbr_set - used_set}
    
    # === REPORT: MISSING KEYS ===
    print(f"\n{'=' * 70}")
    print(f"❌ MISSING KEYS ({len(missing)} keys used in code but NOT in pt-BR)")
    print(f"{'=' * 70}")
    
    if missing:
        # Group by top-level namespace
        by_namespace = defaultdict(dict)
        for key, files in sorted(missing.items()):
            namespace = key.split(".")[0]
            by_namespace[namespace][key] = files[0]  # first file as reference
        
        for ns in sorted(by_namespace.keys()):
            print(f"\n  [{ns}]")
            for key, first_file in sorted(by_namespace[ns].items()):
                print(f"    ✗ {key}")
                print(f"        → {first_file}")
    else:
        print("  🎉 No missing keys found!")
    
    # === REPORT: UNUSED KEYS ===
    print(f"\n{'=' * 70}")
    print(f"⚠️  POTENTIALLY UNUSED KEYS ({len(unused)} keys in pt-BR not found in source)")
    print(f"{'=' * 70}")
    
    if unused:
        by_file = defaultdict(list)
        for key, info in sorted(unused.items()):
            by_file[info["file"]].append(key)
        
        for fname in sorted(by_file.keys()):
            print(f"\n  [{fname}]")
            for key in sorted(by_file[fname]):
                print(f"    ~ {key}")
    else:
        print("  🎉 No unused keys found!")
    
    # === SUMMARY ===
    print(f"\n{'=' * 70}")
    print(f"📊 SUMMARY")
    print(f"{'=' * 70}")
    print(f"  pt-BR keys:      {len(ptbr_keys):>5}")
    print(f"  Used in source:  {len(used_keys):>5}")
    print(f"  Missing (❌):    {len(missing):>5}")
    print(f"  Unused (⚠️):     {len(unused):>5}")
    print()
    
    if missing:
        print("\n💡 JSON snippet to add missing keys (grouped by file):")
        print("-" * 70)
        
        # Group missing keys by their first segment to suggest which JSON file
        by_file_suggestion = defaultdict(dict)
        for key in sorted(missing.keys()):
            parts = key.split(".")
            ns = parts[0]
            # Suggest file name
            file_map = {
                "common": "common.json",
                "trades": "trades.json",
                "settings": "settings.json",
                "dashboard": "dashboard.json",
                "auth": "auth.json",
                "analysis": "analysis.json",
                "risk": "risk.json",
                "finance": "finance.json",
                "fiscal": "fiscal.json",
                "reports": "reports.json",
                "strategies": "strategies.json",
                "navigation": "navigation.json",
                "onboarding": "onboarding.json",
                "ai": "ai.json",
                "accounts": "accounts.json",
            }
            suggested_file = file_map.get(ns, f"{ns}.json")
            
            # Build nested dict for pretty JSON
            nested = by_file_suggestion[suggested_file]
            current = nested
            for i, part in enumerate(parts[1:], 1):
                if i == len(parts) - 1:
                    current.setdefault(part, f"TODO: {key}")
                else:
                    current = current.setdefault(part, {})
        
        for fname, nested in sorted(by_file_suggestion.items()):
            print(f"\n  // Add to: src/lib/i18n/locales/pt-BR/{fname}")
            print(json.dumps(nested, indent=2, ensure_ascii=False))
    
    return len(missing)


if __name__ == "__main__":
    missing_count = main()
    sys.exit(0 if missing_count == 0 else 1)

File: src/lib/i18n/audit_i18n.mjs
/**
 * TraderLogPro — i18n Full Audit
 * Finds hardcoded strings AND missing $t() keys in .svelte files
 * Run: node src/lib/i18n/audit_i18n.mjs
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, relative } from 'path';

const BASE_DIR = 'c:/PROJETOS/TraderLogPro';
const SRC_DIR = join(BASE_DIR, 'src');
const LOCALE_DIR = join(BASE_DIR, 'src/lib/i18n/locales/pt-BR');

const EXCLUDE_DIRS = new Set(['node_modules', '.svelte-kit', 'build', 'dist', '__tests__', '.git']);
const EXCLUDE_FILES = new Set(['audit_i18n.mjs', 'validate_i18n.py', 'index.ts']);

// ─── Load all pt-BR keys (flattened) ────────────────────────────────────────
function flattenKeys(obj, prefix = '') {
  const result = {};
  for (const [k, v] of Object.entries(obj)) {
    const full = prefix ? `${prefix}.${k}` : k;
    if (typeof v === 'object' && v !== null && !Array.isArray(v)) {
      Object.assign(result, flattenKeys(v, full));
    } else {
      result[full] = v;
    }
  }
  return result;
}

function loadPtBrKeys() {
  const files = readdirSync(LOCALE_DIR).filter(f => f.endsWith('.json'));
  const allKeys = {};
  const byFile = {};
  for (const file of files) {
    const raw = JSON.parse(readFileSync(join(LOCALE_DIR, file), 'utf8'));
    const flat = flattenKeys(raw);
    byFile[file] = flat;
    for (const [k, v] of Object.entries(flat)) allKeys[k] = { value: v, file };
  }
  return { allKeys, byFile };
}

// ─── Walk source files ───────────────────────────────────────────────────────
function walkDir(dir, result = []) {
  for (const entry of readdirSync(dir)) {
    if (EXCLUDE_DIRS.has(entry)) continue;
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) walkDir(full, result);
    else if (entry.endsWith('.svelte') || (entry.endsWith('.ts') && !entry.includes('.test.')))
      result.push(full);
  }
  return result;
}

// ─── Extract $t() keys ────────────────────────────────────────────────────────
function extractTKeys(content) {
  const keys = new Set();
  // Match $t('key') and $t("key") — static keys only
  const re = /\$t\(\s*['"]([^'"{}]+)['"]/g;
  let m;
  while ((m = re.exec(content)) !== null) keys.add(m[1]);
  return keys;
}

// ─── Find hardcoded text in Svelte templates ─────────────────────────────────
function findHardcodedStrings(content, filepath) {
  const hits = [];
  const lines = content.split('\n');
  
  // Patterns that signal hardcoded text (outside of script tags)
  // We skip script blocks and style blocks
  let inScript = false, inStyle = false;
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineNum = i + 1;
    const trimmed = line.trim();
    
    if (trimmed.startsWith('<script')) { inScript = true; continue; }
    if (trimmed.startsWith('</script>')) { inScript = false; continue; }
    if (trimmed.startsWith('<style')) { inStyle = true; continue; }
    if (trimmed.startsWith('</style>')) { inStyle = false; continue; }
    if (inScript || inStyle) continue;
    
    // Skip empty lines, HTML-only lines, Svelte control flow
    if (!trimmed || trimmed.startsWith('{#') || trimmed.startsWith('{/') || 
        trimmed.startsWith('{:') || trimmed.startsWith('<!--')) continue;
    
    // Check for text content between tags (not $t, not expressions)
    // Pattern: >SomeText< where text is 3+ chars, not pure numbers/symbols
    const textContent = trimmed.match(/>([^<{]{3,})</g);
    if (textContent) {
      for (const match of textContent) {
        const text = match.replace(/^>/, '').replace(/<$/, '').trim();
        if (
          text.length >= 3 &&
          !/^\d+[\d.,\s%]*$/.test(text) &&  // skip pure numbers
          !/^[^a-zA-ZÀ-ÿ]+$/.test(text) &&  // skip no-letter strings
          !text.includes('{') &&              // skip expressions
          !text.includes('$t') &&            // skip already translated
          !/^(true|false|null|undefined)$/.test(text)
        ) {
          hits.push({ line: lineNum, text, context: 'text_content', raw: trimmed });
        }
      }
    }
    
    // Check title= and placeholder= attributes with hardcoded strings
    const attrMatches = trimmed.matchAll(/(?:title|placeholder|label|aria-label)="([^"]{3,})"/g);
    for (const attr of attrMatches) {
      const text = attr[1];
      if (!text.includes('{') && !/^\d+$/.test(text)) {
        hits.push({ line: lineNum, text, context: 'attribute', raw: trimmed });
      }
    }
  }
  
  return hits;
}

// ─── Main ────────────────────────────────────────────────────────────────────
const { allKeys: ptBrKeys } = loadPtBrKeys();
const files = walkDir(SRC_DIR);

const report = {
  missingKeys: {},    // $t('key') used but not in pt-BR
  hardcoded: {},      // text content not using $t()
};

const usedKeys = new Set();

for (const filepath of files) {
  const content = readFileSync(filepath, 'utf8');
  const rel = relative(BASE_DIR, filepath).replace(/\\/g, '/');

  // Skip i18n files
  if (rel.includes('i18n')) continue;
  
  // Collect $t() keys
  const tKeys = extractTKeys(content);
  for (const k of tKeys) usedKeys.add(k);
  
  // Find missing keys
  const missing = [...tKeys].filter(k => !ptBrKeys[k]);
  if (missing.length > 0) {
    report.missingKeys[rel] = missing.sort();
  }
  
  // Find hardcoded strings (only in .svelte files for template analysis)
  if (filepath.endsWith('.svelte')) {
    const hardcoded = findHardcodedStrings(content, filepath);
    if (hardcoded.length > 0) {
      report.hardcoded[rel] = hardcoded;
    }
  }
}

// ─── Output ──────────────────────────────────────────────────────────────────
console.log('='.repeat(72));
console.log('TraderLogPro — i18n FULL AUDIT REPORT');
console.log('='.repeat(72));

// SECTION 1: Missing $t() keys
const totalMissing = Object.values(report.missingKeys).flat().length;
console.log(`\n❌ MISSING KEYS IN PT-BR (${totalMissing} keys across ${Object.keys(report.missingKeys).length} files)`);
console.log('-'.repeat(72));

// Group by route/component section
const bySection = {};
for (const [file, keys] of Object.entries(report.missingKeys)) {
  const section = file.split('/').slice(0, 4).join('/');
  if (!bySection[section]) bySection[section] = {};
  bySection[section][file] = keys;
}

for (const [section, files] of Object.entries(bySection).sort()) {
  console.log(`\n  📁 ${section}`);
  for (const [file, keys] of Object.entries(files)) {
    console.log(`     ${file.split('/').pop()}`);
    for (const k of keys) console.log(`       ✗ ${k}`);
  }
}

// SECTION 2: Hardcoded strings (grouped by route)
const totalHardcoded = Object.values(report.hardcoded).flat().length;
console.log(`\n\n⚠️  HARDCODED STRINGS IN TEMPLATES (${totalHardcoded} occurrences)`);
console.log('-'.repeat(72));

const svelteRoutes = Object.entries(report.hardcoded)
  .filter(([f]) => f.includes('routes'))
  .sort(([a], [b]) => a.localeCompare(b));

const svelteComponents = Object.entries(report.hardcoded)
  .filter(([f]) => f.includes('components'))
  .sort(([a], [b]) => a.localeCompare(b));

console.log('\n  📁 ROUTES (Pages):');
for (const [file, hits] of svelteRoutes) {
  const uniqueTexts = [...new Set(hits.map(h => h.text))];
  console.log(`     ${file}`);
  for (const t of uniqueTexts.slice(0, 8)) {
    console.log(`       ~ "${t}"`);
  }
  if (uniqueTexts.length > 8) console.log(`       ... +${uniqueTexts.length - 8} more`);
}

console.log('\n  📁 COMPONENTS:');
for (const [file, hits] of svelteComponents) {
  const uniqueTexts = [...new Set(hits.map(h => h.text))];
  console.log(`     ${file}`);
  for (const t of uniqueTexts.slice(0, 5)) {
    console.log(`       ~ "${t}"`);
  }
  if (uniqueTexts.length > 5) console.log(`       ... +${uniqueTexts.length - 5} more`);
}

// SECTION 3: Summary
console.log('\n' + '='.repeat(72));
console.log('📊 SUMMARY');
console.log('='.repeat(72));
console.log(`  pt-BR keys defined:     ${Object.keys(ptBrKeys).length}`);
console.log(`  $t() keys in source:    ${usedKeys.size}`);
console.log(`  Missing from pt-BR:     ${totalMissing}`);
console.log(`  Hardcoded strings:      ${totalHardcoded}`);
console.log(`  Files with issues:      ${Object.keys(report.missingKeys).length + Object.keys(report.hardcoded).length}`);
console.log();
